<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Introduction to reptiledbr: Accessing The Reptile Database in R</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to reptiledbr: Accessing The
Reptile Database in R</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(reptiledbr)</span></code></pre></div>
<p><code>reptiledbr</code> is an R package that provides an interface to
The Reptile Database, allowing researchers and reptile enthusiasts to
easily access taxonomic information, distribution data, and other
details about reptile species. This tutorial will walk you through the
basic usage of the package.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>You can install the <code>reptiledbr</code> package from GitHub
using:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># install.packages(&quot;pak&quot;)</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>pak<span class="sc">::</span><span class="fu">pak</span>(<span class="st">&quot;PaulESantos/reptiledbr&quot;</span>)</span></code></pre></div>
</div>
<div id="basic-usage" class="section level2">
<h2>Basic Usage</h2>
<div id="loading-the-package" class="section level3">
<h3>1. Loading the package</h3>
<p>Let’s start by loading the package:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(reptiledbr)</span></code></pre></div>
</div>
<div id="retrieving-data-for-specific-species" class="section level3">
<h3>2. Retrieving data for specific species</h3>
<p>The main function of the package is
<code>get_reptiledb_data()</code>, which allows you to search for one or
more species:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Create a list of species to search</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>species_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="st">&quot;Lachesis muta&quot;</span>,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="st">&quot;Python bivittatus&quot;</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="st">&quot;Crotalus atrox&quot;</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>)</span></code></pre></div>
<p>This function will connect to The Reptile Database, search for each
species, and download the available information. During the process,
you’ll see progress messages like:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Get data for these species</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>reptile_data <span class="ot">&lt;-</span> <span class="fu">get_reptiledb_data</span>(species_list)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; Starting search for 3 species...</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt; Maximum wait time per request: 10 seconds</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt; Processing: Lachesis muta</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Lachesis muta</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; Processing: Python bivittatus</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Python bivittatus</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; Processing: Crotalus atrox</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Crotalus atrox</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; Search summary:</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt;  Species found with data: 3</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;  Species not found: 0</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt;  Total errors: 0</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span></code></pre></div>
</div>
<div id="understanding-the-returned-data" class="section level3">
<h3>3. Understanding the returned data</h3>
<p>The <code>get_reptiledb_data()</code> function returns a tibble with
columns containing both metadata about the search and the actual data
retrieved:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>reptile_data</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 3 × 7</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt;   input_name        genus    species    url        status error_message data    </span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt;         &lt;list&gt;  </span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; 1 Lachesis muta     Lachesis muta       http://re… succe… &lt;NA&gt;          &lt;tibble&gt;</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; 2 Python bivittatus Python   bivittatus http://re… succe… &lt;NA&gt;          &lt;tibble&gt;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; 3 Crotalus atrox    Crotalus atrox      http://re… succe… &lt;NA&gt;          &lt;tibble&gt;</span></span></code></pre></div>
<p>The <code>data</code> column contains nested tibbles with all the
information scraped from each species page.</p>
</div>
<div id="formatting-data-with-specialized-functions" class="section level3">
<h3>4. Formatting data with specialized functions</h3>
<p><code>reptiledbr</code> provides several specialized functions to
extract and format specific types of information:</p>
<div id="synonyms" class="section level4">
<h4>Synonyms</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">format_synonyms</span>(reptile_data)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 78 × 4</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt;    input_name    genus    species synonym                                       </span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;                                         </span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta Lachesis muta    Crotalus mutus LINNAEUS 1766: 373             </span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta Lachesis muta    Coluber crotalinus GMELIN 1789: 1094          </span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta Lachesis muta    Scytale catenatus LATREILLE in SONNINI and LA…</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt;  4 Lachesis muta Lachesis muta    Scytale ammodytes LATREILLE in SONNINI and LA…</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt;  5 Lachesis muta Lachesis muta    Coluber alecto SHAW 1802: 405                 </span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt;  6 Lachesis muta Lachesis muta    Lachesis mutus — DAUDIN 1803: 351             </span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt;  7 Lachesis muta Lachesis muta    Lachesis ater — DAUDIN 1803: 354              </span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt;  8 Lachesis muta Lachesis muta    Trigonocephalus ammodytes — OPPEL 1811: 390   </span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt;  9 Lachesis muta Lachesis muta    Cophias crotalinus — MERREM 1820: 144         </span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#&gt; 10 Lachesis muta Lachesis muta    Trigonocephalus crotalinus — SCHINZ 1822: 144 </span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 68 more rows</span></span></code></pre></div>
</div>
<div id="distribution-information" class="section level4">
<h4>Distribution information</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">format_distribution</span>(reptile_data)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 11 × 4</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#&gt;    input_name        genus    species    distribution                           </span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                  </span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta     Lachesis muta       &quot;Colombia, E Ecuador, Brazil (Minas Ge…</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta     Lachesis muta       &quot;rhombeata: Brazil (Alagoas, Bahia, Ri…</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta     Lachesis muta       &quot;Type locality: Suriname; restricted t…</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;  4 Python bivittatus Python   bivittatus &quot;SE Nepal, India (Assam, Tripura, Sikk…</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt;  5 Python bivittatus Python   bivittatus &quot;Introduced to Florida (USA)&quot;          </span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;  6 Python bivittatus Python   bivittatus &quot;progschai: Sulawesi; Type locality: S…</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;  7 Python bivittatus Python   bivittatus &quot;Type locality: Java (designated by ME…</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt;  8 Crotalus atrox    Crotalus atrox      &quot;USA (SE California, S Nevada [A Heind…</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt;  9 Crotalus atrox    Crotalus atrox      &quot;Mexico (Mexico [HR 35: 190], Baja Cal…</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; 10 Crotalus atrox    Crotalus atrox      &quot;Type locality: \&quot;Indianola\&quot; [Indiano…</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt; 11 Crotalus atrox    Crotalus atrox      &quot;tortugensis: Mexico (Isla Tortuga in …</span></span></code></pre></div>
</div>
<div id="higher-taxonomic-classifications" class="section level4">
<h4>Higher taxonomic classifications</h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">format_higher_taxa</span>(reptile_data)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 20 × 4</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt;    input_name        genus    species    taxon            </span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;            </span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta     Lachesis muta       Viperidae        </span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta     Lachesis muta       Crotalinae       </span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta     Lachesis muta       Colubroidea      </span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;  4 Lachesis muta     Lachesis muta       Caenophidia      </span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;  5 Lachesis muta     Lachesis muta       Alethinophidia   </span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;  6 Lachesis muta     Lachesis muta       Serpentes        </span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt;  7 Lachesis muta     Lachesis muta       Squamata (snakes)</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt;  8 Python bivittatus Python   bivittatus Pythonidae       </span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt;  9 Python bivittatus Python   bivittatus Henophidia       </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; 10 Python bivittatus Python   bivittatus Pythonoidea      </span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt; 11 Python bivittatus Python   bivittatus Alethinophidia   </span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#&gt; 12 Python bivittatus Python   bivittatus Serpentes        </span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#&gt; 13 Python bivittatus Python   bivittatus Squamata (snakes)</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#&gt; 14 Crotalus atrox    Crotalus atrox      Viperidae        </span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co">#&gt; 15 Crotalus atrox    Crotalus atrox      Crotalinae       </span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co">#&gt; 16 Crotalus atrox    Crotalus atrox      Colubroidea      </span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt; 17 Crotalus atrox    Crotalus atrox      Caenophidia      </span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#&gt; 18 Crotalus atrox    Crotalus atrox      Alethinophidia   </span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="co">#&gt; 19 Crotalus atrox    Crotalus atrox      Serpentes        </span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="co">#&gt; 20 Crotalus atrox    Crotalus atrox      Squamata (snakes)</span></span></code></pre></div>
</div>
<div id="subspecies-information" class="section level4">
<h4>Subspecies information</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">format_subspecies</span>(reptile_data)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 2 × 4</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co">#&gt;   input_name        genus  species    subspecies                                </span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;                                     </span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; 1 Python bivittatus Python bivittatus Python bivittatus progschai JACOBS, AULYI…</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co">#&gt; 2 Python bivittatus Python bivittatus Python bivittatus bivittatus KUHL 1820</span></span></code></pre></div>
</div>
<div id="common-names" class="section level4">
<h4>Common names</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">format_common_names</span>(reptile_data)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 13 × 4</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt;    input_name        genus    species    common_name                            </span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                  </span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta     Lachesis muta       E: South American Bushmaster           </span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta     Lachesis muta       G: Südamerikanischer Buschmeister      </span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta     Lachesis muta       rhombeata: Atlantic Forest bushmaster  </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt;  4 Lachesis muta     Lachesis muta       NL: Bosmeester                         </span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt;  5 Lachesis muta     Lachesis muta       Portuguese: Bico-de-Jaca, Cobra-Topete…</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt;  6 Lachesis muta     Lachesis muta       S: Cuaima; Guayma concha de piña       </span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt;  7 Python bivittatus Python   bivittatus E: Burmese Python                      </span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt;  8 Python bivittatus Python   bivittatus G: Dunkler Tigerpython                 </span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt;  9 Python bivittatus Python   bivittatus Chinese: 蟒                            </span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; 10 Crotalus atrox    Crotalus atrox      E: Western Diamond-backed Rattlesnake  </span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt; 11 Crotalus atrox    Crotalus atrox      G: Texas-Klapperschlange, Westliche Di…</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt; 12 Crotalus atrox    Crotalus atrox      E: Tortuga Island Rattlesnake (tortuge…</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co">#&gt; 13 Crotalus atrox    Crotalus atrox      S: Cascabel de Diamantes</span></span></code></pre></div>
</div>
<div id="reproduction-information" class="section level4">
<h4>Reproduction information</h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">format_reproduction</span>(reptile_data)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 3 × 4</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt;   input_name        genus    species    reproduction_detail                     </span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                   </span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; 1 Lachesis muta     Lachesis muta       oviparous; one of the few snakes that s…</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt; 2 Python bivittatus Python   bivittatus GROOT et al. (2003) and KUHN &amp; SCHMIDT …</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt; 3 Crotalus atrox    Crotalus atrox      ovoviviparous. C. atrox and C. horridus…</span></span></code></pre></div>
</div>
<div id="type-specimens" class="section level4">
<h4>Type specimens</h4>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">format_types</span>(reptile_data)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 4</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co">#&gt;   input_name        genus    species    type_detail                             </span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                   </span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co">#&gt; 1 Lachesis muta     Lachesis muta       &quot;Holotype: NRM = NHRM. According to And…</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt; 2 Lachesis muta     Lachesis muta       &quot;Holotype: ZSM, uncatalogued specimen(s…</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt; 3 Python bivittatus Python   bivittatus &quot;Holotype: iconotype: Plate in SEBA 173…</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt; 4 Python bivittatus Python   bivittatus &quot;Holotype: ZFMK 87481, subadult male [p…</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#&gt; 5 Crotalus atrox    Crotalus atrox      &quot;Holotype: USNM 7761&quot;                   </span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co">#&gt; 6 Crotalus atrox    Crotalus atrox      &quot;Holotype: CAS 50515 [tortugensis]&quot;</span></span></code></pre></div>
</div>
<div id="diagnosis-information" class="section level4">
<h4>Diagnosis information</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">format_diagnosis</span>(reptile_data)</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 5 × 4</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co">#&gt;   input_name        genus    species    diagnosis_detail                        </span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                   </span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co">#&gt; 1 Lachesis muta     Lachesis muta       Diagnosis: Distinguished from L. stenop…</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="co">#&gt; 2 Lachesis muta     Lachesis muta       Additional details (3041 characters) ar…</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="co">#&gt; 3 Python bivittatus Python   bivittatus Diagnosis: Large serpents; labials sepa…</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co">#&gt; 4 Python bivittatus Python   bivittatus Additional details (2449 characters) ar…</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="co">#&gt; 5 Crotalus atrox    Crotalus atrox      Additional details, e.g. a detailed des…</span></span></code></pre></div>
</div>
<div id="comments" class="section level4">
<h4>Comments</h4>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">format_comments</span>(reptile_data)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 15 × 4</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt;    input_name        genus    species    comment_detail                         </span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                                  </span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta     Lachesis muta       &quot;Venomous!&quot;                            </span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta     Lachesis muta       &quot;Synonymy: Lachesis muta has been spli…</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta     Lachesis muta       &quot;Type species: Lachesis mutus is the t…</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt;  4 Lachesis muta     Lachesis muta       &quot;Distribution: see maps in Barrio-Amor…</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt;  5 Python bivittatus Python   bivittatus &quot;Subspecies: This species has been con…</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt;  6 Python bivittatus Python   bivittatus &quot;Hybridization: \&quot;The evidence for the…</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt;  7 Python bivittatus Python   bivittatus &quot;Distribution: Records from Sumatra an…</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt;  8 Python bivittatus Python   bivittatus &quot;DNA barcodes suggested multiple speci…</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt;  9 Python bivittatus Python   bivittatus &quot;Genome: Python (molurus) bivittatus w…</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt; 10 Crotalus atrox    Crotalus atrox      &quot;Venomous! Crotalus atrox is responsib…</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#&gt; 11 Crotalus atrox    Crotalus atrox      &quot;Nomenclature: Hoser’s 2009 classifica…</span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="co">#&gt; 12 Crotalus atrox    Crotalus atrox      &quot;Synonymy: AMARAL 1929 synonymed C. to…</span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="co">#&gt; 13 Crotalus atrox    Crotalus atrox      &quot;Similar species: C. scutulatus.&quot;      </span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="co">#&gt; 14 Crotalus atrox    Crotalus atrox      &quot;Love (2013) reports melanism in C. at…</span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="co">#&gt; 15 Crotalus atrox    Crotalus atrox      &quot;Habitat: usually terrestrial, occasio…</span></span></code></pre></div>
</div>
<div id="etymology" class="section level4">
<h4>Etymology</h4>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">format_etymology</span>(reptile_data)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 2 × 4</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="co">#&gt;   input_name    genus    species etymology_detail                               </span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;                                          </span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co">#&gt; 1 Lachesis muta Lachesis muta    The specific name muta is from the Latin mutus…</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#&gt; 2 Lachesis muta Lachesis muta    The genus was apparently named after Lachesis,…</span></span></code></pre></div>
</div>
<div id="references" class="section level4">
<h4>References</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">format_references</span>(reptile_data)</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 418 × 4</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co">#&gt;    input_name    genus    species reference                                     </span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;                                         </span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="co">#&gt;  1 Lachesis muta Lachesis muta    Alves, Fátima Q.; AntÃ´nio J. S. ArgÃ´lo, and…</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="co">#&gt;  2 Lachesis muta Lachesis muta    Amaral, A. do 1925. On the Oviparity of Lache…</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co">#&gt;  3 Lachesis muta Lachesis muta    Amaral,A. do 1926. 4.a nota de nomenclatura O…</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="co">#&gt;  4 Lachesis muta Lachesis muta    Andersson, L.G. 1899. Catalogue of Linnean ty…</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a><span class="co">#&gt;  5 Lachesis muta Lachesis muta    ANDRADE LIMA, J.H., DIAS, E.G., COSTA, R.D.L.…</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a><span class="co">#&gt;  6 Lachesis muta Lachesis muta    Araujo Filho, João Antonio de, Cícero Ricardo…</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co">#&gt;  7 Lachesis muta Lachesis muta    Arteaga, A.; Bustamante, L.; Vieira, J. 2024.…</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a><span class="co">#&gt;  8 Lachesis muta Lachesis muta    Avila-Pires, Teresa C.S.; Kleiton R. Alves-Si…</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a><span class="co">#&gt;  9 Lachesis muta Lachesis muta    Barbour, Thomas 1930. The bushmaster in the C…</span></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="co">#&gt; 10 Lachesis muta Lachesis muta    Barrio-Amorós, Cesar L.; Greivin Corrales, Sy…</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 408 more rows</span></span></code></pre></div>
</div>
</div>
</div>
<div id="handling-special-cases" class="section level2">
<h2>Handling Special Cases</h2>
<p>The package provides proper error handling for various cases:</p>
<div id="trinomial-names-subspecies" class="section level3">
<h3>1. Trinomial names (subspecies)</h3>
<p>If you search for a trinomial name (genus + species + subspecies),
the package will issue a warning and skip the search:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>species_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>  <span class="st">&quot;Lachesis muta&quot;</span>,</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>  <span class="st">&quot;Bothrops atrox insularis&quot;</span>  <span class="co"># Trinomial name</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>)</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>reptile_data <span class="ot">&lt;-</span> <span class="fu">get_reptiledb_data</span>(species_list)</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="co">#&gt; Starting search for 2 species...</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co">#&gt; Maximum wait time per request: 10 seconds</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="co">#&gt; Processing: Lachesis muta</span></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Lachesis muta</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="co">#&gt; Processing: Bothrops atrox insularis</span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: Trinomial name detected: Bothrops atrox insularis</span></span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a><span class="co">#&gt;    The database only supports binomial names (genus and species).</span></span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a><span class="co">#&gt;    No search will be performed for trinomial names.</span></span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a><span class="co">#&gt; Search summary:</span></span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a><span class="co">#&gt;  Species found with data: 1</span></span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a><span class="co">#&gt;  Species not found: 0</span></span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a><span class="co">#&gt;  Total errors: 1</span></span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a><span class="co">#&gt;    - Trinomial names (not supported): 1</span></span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span></code></pre></div>
</div>
<div id="incomplete-species-names" class="section level3">
<h3>2. Incomplete species names</h3>
<p>The package can also handle incomplete species names, but it will
warn you if no data is found:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>species_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>  <span class="st">&quot;Lachesis muta&quot;</span>,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="st">&quot;Lachesis sp&quot;</span>  <span class="co"># Incomplete species name</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>reptile_data <span class="ot">&lt;-</span> <span class="fu">get_reptiledb_data</span>(species_list)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="co">#&gt; Starting search for 2 species...</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co">#&gt; Maximum wait time per request: 10 seconds</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="co">#&gt; Processing: Lachesis muta</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Lachesis muta</span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="co">#&gt; Processing: Lachesis sp</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a><span class="co">#&gt; SUCCESS: Data found for Lachesis sp</span></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: No data table found for Lachesis sp</span></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a><span class="co">#&gt;    This is considered an error and the URL will be removed.</span></span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a><span class="co">#&gt; Search summary:</span></span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a><span class="co">#&gt;  Species found with data: 1</span></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a><span class="co">#&gt;  Species not found: 0</span></span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a><span class="co">#&gt;  Total errors: 1</span></span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a><span class="co">#&gt;    - Pages without data tables: 1</span></span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a><span class="co">#&gt; ------------------------------------------------------------------------</span></span></code></pre></div>
</div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p><code>reptiledbr</code> provides an easy and structured way to access
reptile species information from <a href="http://www.reptile-database.org/">The Reptile Database</a>. The
package handles many common edge cases and provides specialized
functions to extract and format specific types of information. This
makes it a valuable tool for herpetologists, ecologists, and other
researchers working with reptile data.</p>
<p>For more information about the package, including advanced usage and
complete function documentation, please refer to the package’s <a href="https://github.com/PaulESantos/reptiledbr">documentation</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
