<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>cnv</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">cnv</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(oncoPredict)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#This script provides an example of how to download cnv (copy number variation) data from the GDC database for GBM #(glioblastoma), how to apply the map_cnv() function to that data to map cnvs to genes, and how to test the drugs in your drug #response dataset to each cnv to identify biomarkers that enrich for drug response.</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#First, download CNV data for your cancer of interest from GDC database. The cnv data will be exported to your #working directory as cnv.txt </span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#This code will export the CNV data into a text file called, &#39;cnv.txt&#39;, containing a table with colnames() &#39;Sample&#39;, &#39;Chromosome&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Num Probes&#39;, &#39;Segment_Mean&#39;</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#The genome of reference is hg19.</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#query.gbm.nocnv&lt;-GDCquery(project = &quot;TCGA-GBM&quot;,</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#                          data.category = &quot;Copy number variation&quot;,</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#                          legacy = TRUE,</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#                          file.type = &quot;nocnv_hg19.seg&quot;,</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#                          sample.type = c(&quot;Primary Tumor&quot;))</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#patient_total&lt;-nrow((query.gbm.nocnv$results)[[1]]) #The total number of patients GDC has CNV data for</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#query.gbm.nocnv$results[[1]]&lt;-query.gbm.nocnv$results[[1]][1:patient_total,]</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#GDCdownload(query.gbm.nocnv, files.per.chunk = 100)</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co">#gbm.nocnv&lt;-GDCprepare(query.gbm.nocnv, save = TRUE, save.filename = &quot;GBMnocnvhg19.rda&quot;)</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co">#write.table(gbm.nocnv, file=&#39;cnv.txt&#39;)</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co">#Second, apply map_cnv() to map cnv data to genes. The mapped cnv data will be exported to your working directory as map.RData</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co">#The mapping is accomplished by intersecting the gene with the overlapping CNV level. If the gene isn&#39;t fully #captured by the CNV, an NA will be assigned.</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co">#Determine the parameters of the map_cnv() function.</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a>Cnvs<span class="ot">&lt;-</span><span class="fu">read.table</span>(<span class="st">&#39;cnv.txt&#39;</span>, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co">#Third, apply idwas() to test each cnv and each drug. The p-values and beta-values for each test will be exported to #your working directory as CnvTestOutput_pVals.txt and CnvTestOutput_betas.txt</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="co">#Determine the parameters of the idwas() function...</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="co">#Set the drug_prediction parameter.</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="co">#Make sure rownames() are samples, and colnames() are drugs. Also make sure this data is a data frame.</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>drug_prediction<span class="ot">&lt;-</span><span class="fu">t</span>(<span class="fu">as.data.frame</span>(<span class="fu">read.table</span>(<span class="st">&#39;DrugPredictions.txt&#39;</span>, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="dv">1</span>)))</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="co">#dim(drug_prediction) #165 198</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co">#In this example, I had to replace the &#39;.&#39; in the names of these TCGA samples with &#39;-&#39; so that they are of the same form as samples in the cnv  data (you may not have to do this).</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="fu">rownames</span>(drug_prediction)<span class="ot">&lt;-</span><span class="fu">gsub</span>(<span class="st">&quot;.&quot;</span>, <span class="st">&quot;-&quot;</span>, <span class="fu">rownames</span>(drug_prediction), <span class="at">fixed=</span>T)</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="co">#Make sure the sample identifiers in the &#39;drug prediction&#39; data are of similar form as the sample identifiers in the &#39;data&#39; parameter.</span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a>rows<span class="ot">=</span><span class="fu">rownames</span>(drug_prediction)</span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="fu">rownames</span>(drug_prediction)<span class="ot">&lt;-</span><span class="fu">substring</span>(rows, <span class="dv">3</span>, <span class="fu">nchar</span>(rows))</span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a>drug_prediction<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(drug_prediction)</span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="co">#Determine the number of samples you want the CNVs to be amplified in. The default is 10.</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">10</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="co">#Indicate whether or not you would like to test cnv data. If TRUE, you will test cnv data. If FALSE, you will test mutation data.</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a>cnv<span class="ot">=</span><span class="cn">TRUE</span></span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a>wd<span class="ot">&lt;-</span><span class="fu">tempdir</span>()</span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a>savedir<span class="ot">&lt;-</span><span class="fu">setwd</span>(wd)</span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="co">#Apply map_cnv()</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="co">#This function produces the file map.RData, which stores the object &#39;theCnvQuantVecList_mat&#39;</span></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="fu">map_cnv</span>(<span class="at">Cnvs=</span>Cnvs)</span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="co">#&gt;   403 genes were dropped because they have exons located on both strands</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a><span class="co">#&gt;   of the same reference sequence or on more than one reference sequence,</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a><span class="co">#&gt;   so cannot be represented by a single genomic range.</span></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a><span class="co">#&gt;   Use &#39;single.strand.genes.only=FALSE&#39; to get all the genes in a</span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a><span class="co">#&gt;   GRangesList object, or use suppressMessages() to suppress this message.</span></span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a><span class="co">#Set the data parameter.</span></span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&#39;map.RData&#39;</span>) <span class="co">#This loads the object &#39;theCnvQuantVecList_mat&#39;, which was obtained using map_cnv()</span></span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a><span class="co">#Make sure this data is a data frame and that colnames() are samples.</span></span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a>data<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">t</span>(theCnvQuantVecList_mat))</span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a>samps<span class="ot">&lt;-</span><span class="fu">colnames</span>(data)</span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="fu">colnames</span>(data)<span class="ot">&lt;-</span><span class="fu">substr</span>(samps,<span class="dv">1</span>,<span class="fu">nchar</span>(samps)<span class="sc">-</span><span class="dv">12</span>)</span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a></span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a><span class="co">#Apply idwas()</span></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a><span class="fu">idwas</span>(<span class="at">drug_prediction=</span>drug_prediction, <span class="at">data=</span>data, <span class="at">n=</span>n, <span class="at">cnv=</span>cnv)</span></code></pre></div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
