--- ../src/swipl-devel/packages/clib/process.c	2025-03-17 13:25:06.993245104 +0100
+++ ../src/swipl-devel/packages/clib/process.c	2025-03-17 15:00:06.758897819 +0100
@@ -1809,15 +1809,15 @@
 
   blockSignal(SIGPROF, &set);
 
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-  if ( method == PCREATE_VFORK )
-    pid = vfork();				/* As long as it is there we use it */
-  else
+// prag_ma GCC diagnostic push
+// prag_ma GCC diagnostic ignored "-Wdeprecated-declarations"
+//   if ( method == PCREATE_VFORK )
+//     pid = vfork();				/* As long as it is there we use it */
+//   else
     pid = fork();
   if ( pid != 0 )				/* parent */
     restoreSignals(&set);
-#pragma GCC diagnostic pop
+// prag_ma GCC diagnostic pop
 
   if ( pid == 0 )				/* child */
   { int fd;
