--- ../src/swipl/packages/cpp/test_cpp.cpp	2025-03-20 19:12:10.575847401 +0100
+++ ../src/swipl/packages/cpp/test_cpp.cpp	2025-03-20 20:52:33.052384566 +0100
@@ -668,25 +668,25 @@
 // The eq_int64/2 and lt_int64/2 predicates test the deprecated PlTerm::operator==()
 
 #ifdef _MSC_VER
-#pragma warning( push )
-#pragma warning (disable:4996)
+// #pragma warning( push )
+// #pragma warning (disable:4996)
 #else
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+// #pragma GCC diagnostic push
+// #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #endif
 
-PREDICATE(eq_int64, 2)
-{ return A1 == A2.as_int64_t();
-}
+// PREDICATE(eq_int64, 2)
+// { return A1 == A2.as_int64_t();
+// }
 
-PREDICATE(lt_int64, 2)
-{ return A1 < A2.as_int64_t();
-}
+// PREDICATE(lt_int64, 2)
+// { return A1 < A2.as_int64_t();
+// }
 
 #ifdef _MSC_VER
-#pragma warning( pop )
+// #pragma warning( pop )
 #else
-#pragma GCC diagnostic pop
+// #pragma GCC diagnostic pop
 #endif
 
 PREDICATE(get_atom_ex, 2)
