--- ../src/swipl/packages/cpp/test_cpp.pl	2025-03-20 19:13:24.205842130 +0100
+++ ../src/swipl/packages/cpp/test_cpp.pl	2025-03-20 19:10:31.265166927 +0100
@@ -786,6 +786,7 @@
 test(throw, error(resource_error('NO_RESOURCE'))) :-
     throw_resource_error_cpp('NO_RESOURCE').
 
+/*
 test(compare) :-
     eq_int64(1, 1).
 test(compare, fail) :-
@@ -802,6 +803,7 @@
     lt_int64(1, a).
 test(compare, error(type_error(integer,b))) :-
     lt_int64(b, 1).
+*/
 
 test(get_atom, A == abc) :-
     get_atom_ex(abc, A).
