<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Michal Stolarczyk" />

<meta name="date" content="2023-11-21" />

<title>Sample modifiers in pepr: imply and derive</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Sample modifiers in pepr: imply and
derive</h1>
<h4 class="author">Michal Stolarczyk</h4>
<h4 class="date">2023-11-21</h4>



<div id="learn-how-to-combine-implied-and-derived-attributes-in-pepr" class="section level1">
<h1>Learn how to combine implied and derived attributes in
<code>pepr</code></h1>
<p>This vignette will show you how and why to use the derieved
attributes and implied attributes functionalities concurrently of the
<code>pepr</code> package.</p>
<ul>
<li><p>For the basic information about the PEP concept on the <a href="https://pepkit.github.io/">project website</a></p></li>
<li><p>Make sure to study the dedicated <a href="./feature3_derivedAttributes.html">derived attributes</a> and <a href="./feature2_impliedAttributes.html">implied attributes</a>
vignettes prior to reading this one</p></li>
</ul>
<div id="problemgoal" class="section level2">
<h2>Problem/Goal</h2>
<p>While either derived attributes or implied attributes functionalities
alone are often sufficient to efficiently describe your samples in PEP,
the example below demonstrates how to use the derived attributes to
<strong>simplify and unclutter the columns</strong> of the
<code>sample_table.csv</code> file, after implying the attributes for
samples that <strong>follow certain patterns</strong>. The two
functionalities combined provide you with the way of building complex,
yet flexible sample annotation tables effortlessly. Note that the
attributes implication is always performed first - before the attributes
are derived. This means that the newly created attributes (implied ones)
can be used to construct the attributes in the column derivation
process. Please consider the example below for reference:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
sample_name
</th>
<th style="text-align:left;">
organism
</th>
<th style="text-align:right;">
time
</th>
<th style="text-align:left;">
file_path
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
pig_0h
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
data/lab/project/pig_susScr11_untreated.fastq
</td>
</tr>
<tr>
<td style="text-align:left;">
pig_1h
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
data/lab/project/pig_susScr11_treated.fastq
</td>
</tr>
<tr>
<td style="text-align:left;">
frog_0h
</td>
<td style="text-align:left;">
frog
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
data/lab/project/frog_xenTro9_untreated.fastq
</td>
</tr>
<tr>
<td style="text-align:left;">
frog_1h
</td>
<td style="text-align:left;">
frog
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
data/lab/project/frog_xenTro9_treated.fastq
</td>
</tr>
</tbody>
</table>
<hr />
</div>
<div id="solution" class="section level2">
<h2>Solution</h2>
<p>The specification of detailed file paths/names (as presented above)
is cumbersome. In order to make your life easier just find the patterns
that the file names in <code>file_path</code> column of
<code>sample_table.csv</code> follow, imply needed attributes and derive
the file names. This multi step process is orchestrated by the
<code>project_config.yaml</code> file via the
<code>sample_modifiers.derive</code> and
<code>sample_modifiers.imply</code> sections:</p>
<pre><code>   pep_version: 2.0.0
   sample_table: sample_table.csv
   output_dir: $HOME/hello_looper_results
   sample_modifiers:
      derive:
          attributes: file_path
          sources:
              source1: /data/lab/project/{organism}_{genome}_{condition}.fastq
      imply:
              if:
                  organism: pig
              then:
                  genome: susScr11
              if:
                  organism: frog
              then:
                  genome: xenTro9
              if:
                  time: 0
              then:
                  condition: untreated
              if:
                  time: 1
              then:
                  condition: treated</code></pre>
<p>The <code>*_untreated</code> files are clearly associated with the
samples that are labeled with <code>time</code> 0. Therefore the
<code>untreated</code> attribute is implied for the samples which have 0
in the <code>time</code> columns. Similarly, the codes
<code>susScr11</code> and <code>xenTro9</code> are associated with the
attributes in the <code>oragnism</code> column. Therefore, the column
<code>condion</code> that consists of those two codes is implied from
the attributes in the <code>organism</code> column according to the
<code>project_config.yaml</code>.</p>
<p>Letâ€™s introduce a few modifications to the original
<code>sample_table.csv</code> file to imply the attributes
<code>genome</code> and <code>condition</code> and subsequently map the
appropriate data sources from the <code>project_config.yaml</code> with
attributes in the derived column - <code>[file_path]</code>:</p>
<table>
<thead>
<tr>
<th style="text-align:left;">
sample_name
</th>
<th style="text-align:left;">
organism
</th>
<th style="text-align:right;">
time
</th>
<th style="text-align:left;">
file_path
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
pig_0h
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
source1
</td>
</tr>
<tr>
<td style="text-align:left;">
pig_1h
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
source1
</td>
</tr>
<tr>
<td style="text-align:left;">
frog_0h
</td>
<td style="text-align:left;">
frog
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
source1
</td>
</tr>
<tr>
<td style="text-align:left;">
frog_1h
</td>
<td style="text-align:left;">
frog
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
source1
</td>
</tr>
</tbody>
</table>
<hr />
</div>
<div id="code" class="section level2">
<h2>Code</h2>
<p>Load <code>pepr</code> and read in the project metadata by specifying
the path to the <code>project_config.yaml</code>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(pepr)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>projectConfig <span class="ot">=</span> <span class="fu">system.file</span>(</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">&quot;example_peps-&quot;</span>, branch),</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="st">&quot;example_derive_imply&quot;</span>,</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="st">&quot;project_config.yaml&quot;</span>,</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="at">package =</span> <span class="st">&quot;pepr&quot;</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>)</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">Project</span>(projectConfig)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="do">## Loading config file: /tmp/RtmpoymTo9/Rinstb3055bff7/pepr/extdata/example_peps-master/example_derive_imply/project_config.yaml</span></span></code></pre></div>
<p>And inspect it:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">sampleTable</span>(p)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="do">##    sample_name organism time                                      file_path</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="do">## 1:      pig_0h      pig    0 /data/lab/project/pig_susScr11_untreated.fastq</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="do">## 2:      pig_1h      pig    1   /data/lab/project/pig_susScr11_treated.fastq</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="do">## 3:     frog_0h     frog    0 /data/lab/project/frog_xenTro9_untreated.fastq</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="do">## 4:     frog_1h     frog    1   /data/lab/project/frog_xenTro9_treated.fastq</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="do">##      genome condition</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="do">## 1: susScr11 untreated</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="do">## 2: susScr11   treated</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="do">## 3:  xenTro9 untreated</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="do">## 4:  xenTro9   treated</span></span></code></pre></div>
<p>As you can see, the resulting samples are annotated the same way as
if they were read from the original, unwieldy, annotations file
(enriched with the <code>genome</code> and <code>condition</code>
attributes that were implied).</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
